lib/puppet_x/templates/invoke_ps_command.erb

```diff
@@ -0,0 +1,179 @@
+    private bool shouldExit;
+    private int exitCode;
+
+    public PuppetPSHost () {}
+
+    public bool ShouldExit
+    {
+        get { return this.shouldExit; }
+    }
+
+    public int ExitCode
+    {
+        get { return this.exitCode; }
+    }
+
+    public void ResetExitStatus()
+    {
+      this.exitCode = 0;
+      this.shouldExit = false;
+    }
+
+    public override Guid InstanceId
+    {
+        get { return _hostId; }
+    }
+
+    public override string Name
+    {
+        get { return "PuppetPSHost"; }
+    }
+
+    public override Version Version
+    {
+        get { return new Version(1, 0); }
+    }
+
+    public override PSHostUserInterface UI
+    {
+        get { return null; }
+    }
+
+    public override CultureInfo CurrentCulture
+    {
+        get { return Thread.CurrentThread.CurrentCulture; }
+    }
+
+    public override CultureInfo CurrentUICulture
+    {
+        get { return Thread.CurrentThread.CurrentUICulture; }
+    }
+
+    public override void EnterNestedPrompt()
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void ExitNestedPrompt()
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void NotifyBeginApplication()
+    {
+        return;
+    }
+
+    public override void NotifyEndApplication()
+    {
+        return;
+    }
+
+    public override void SetShouldExit(int exitCode)
+    {
+      this.shouldExit = true;
+      this.exitCode = exitCode;
+    }
+}
+"@
+
+Add-Type -TypeDefinition $hostSource -Language CSharp
+
+$dotNetVersion = $PSVersionTable.CLRVersion.Major
+
+$event = [System.Threading.EventWaitHandle]::OpenExisting("<%= output_ready_event_name %>")
+
+if ($runspace -eq $null)
+{
+  # CreateDefault2 requires PS3
+  if ([System.Management.Automation.Runspaces.InitialSessionState].GetMethod('CreateDefault2'))
+  {
+    $sessionState = [System.Management.Automation.Runspaces.InitialSessionState]::CreateDefault2()
+  }
+  else
+  {
+    $sessionState = [System.Management.Automation.Runspaces.InitialSessionState]::CreateDefault()
+  }
+  # $puppetPSHost = [PuppetPSHost]::new()
+  $puppetPSHost = New-Object PuppetPSHost
+  $runspace = [System.Management.Automation.Runspaces.RunspaceFactory]::CreateRunspace($puppetPSHost, $sessionState)
+  $runspace.Open()
+}
+
+$powershell_code = @'
+<%= powershell_code %>
+'@
+
+try
+{
+  $ps = $null
+  $puppetPSHost.ResetExitStatus()
+  $runspace.ResetRunspaceState()
+
+  $ps = [System.Management.Automation.PowerShell]::Create()
+  $ps.Runspace = $runspace
+  [Void]$ps.AddScript($powershell_code)
+
+  $asyncResult = $ps.BeginInvoke()
+
+  if (!$asyncResult.AsyncWaitHandle.WaitOne(<%= timeout_ms %>, $false))
+  {
+    throw "Catastrophic failure: PowerShell module timeout (<%= timeout_ms %> ms) exceeded while executing"
+  }
+
+  $output = $ps.EndInvoke($asyncResult)
+
+@"
+{
+  "stdout": "",
+  "exitcode": $($puppetPSHost.ExitCode),
+  "errormessage": ""
+}
```

Comment:
We have to make our own JSON here, as ConvertTo-JSON exists in PS v3 and above, not in v2. This is WIP